clang/test/SemaSYCL/annotated_arg_or_ptr.cpp

```diff
@@ -0,0 +1,62 @@
+// RUN: %clang_cc1 -fsyntax-only -fsycl-is-device -internal-isystem %S/Inputs -verify %s
+
+// Test diagnostic for nested annotated_arg and annotated_ptr type.
+
+#include "sycl.hpp"
+
+sycl::queue myQueue;
+
+struct MockProperty {};
+
+struct WrappedAnnotatedTypes {
+  // expected-error@+1 3{{Kernel argument of type 'sycl::ext::oneapi::experimental::annotated_arg<int, MockProperty>' cannot be nested}}
+  sycl::ext::oneapi::experimental::annotated_arg<int, MockProperty> AA;
+  // expected-error@+1 3{{Kernel argument of type 'sycl::ext::oneapi::experimental::annotated_ptr<int, MockProperty>' cannot be nested}}
+  sycl::ext::oneapi::experimental::annotated_ptr<int, MockProperty> AP;
+  sycl::accessor<int, 1, sycl::access::mode::read_write> Acc;
+};
+
+struct KernelBase {
+  sycl::ext::oneapi::experimental::annotated_arg<int, MockProperty> BaseAA; // OK
```

Comment:
These 'OK's are a bit a confusing to me.   If 'KernelBase' did not contain the field NestedInBase, would it be okay to use the type 'KernelBase' inside the kernel?

Reply:
You mean declare a field of type KernelBase inside KernelFunctor? I don't think so because that would be considered nested.

Reply:
Annotated types are allowed in kernel base because the following code is not against the spec:

```
struct KernelBase {
  annotated_ptr<int> a;
  KernelBase(annotated_ptr<int> a_) : a(a_) {}
};

struct MyKernel : KernelBase {
  MyKernel(annotated_ptr<int> a_) : KernelBase(a_) {}
  void operator()() const { *a = *a + 1; }
};

int main() {
  sycl::queue q;
  auto ptr = malloc_shared<int>(1, q);
  *ptr = 0;
  q.single_task(MyKernel{ptr}).wait();
  if (*ptr == 1)
    std::cout << "PASSED\n";
  else
    std::cout << "FAILED, output: " << *ptr << "\n";
  sycl::free(ptr, q);
  return 0;
}

```

Reply:
> You mean declare a field of type KernelBase inside KernelFunctor? I don't think so because that would be considered nested.

No, I mean something like this case:

```
struct KernelBase {
  sycl::ext::oneapi::experimental::annotated_arg<int, MockProperty> BaseAA; // OK
  sycl::ext::oneapi::experimental::annotated_ptr<int, MockProperty> BaseAP; // OK
};

int main() {
 KernelBase KB;
  myQueue.submit([&](sycl::handler &h) {
    h.single_task<class kernel_half>(
        [=]() {
          (void)KB;
        });
  });
}
```

Would this be okay?

Reply:
No it won't. Because it is a capture i.e. kernel argument is KernelBase. KernelFunctor is not a capture. It is the SYCL kernel defined as a functor and not a lambda. The argument is the annotated type

Reply:
Okay, and this is what I think is confusing me.   In your original example, what did you mean by the comments 'OK' for the first two KernelBase fields?

Reply:
I meant they don't trigger an error

Reply:
I spoke to @premanandrao briefly offline and we concluded that the current diagnostic `Kernel argument of type %0 cannot be nested` does not accurately capture the issue because the problematic kernel argument is the struct enclosing the annotated type and not the annotated type itself. Something along the lines of `kernel arguments cannot enclose type %0` might make more sense but I do not know if 'enclose' is user friendly. I'm also not a fan of this diagnostic because we're not specifying the 'correct' usage of these types in the diagnostic. I also considered `Variables of type %0 must be captured directly by the kernel` but I do not know if that is confusing for someone who uses kernel functors  @premanandrao please weigh in here. If the other reviewers have a suggestion, please weigh in as well!

Reply:
How about `"Variables of type %0 must be captured directly by the kernel, or declared as a field of the kernel functor"`. I could also look into emitting different diagnostic for kernel defined as a lambda vs kernel functor. 

`"Variables of type %0 must be captured directly by the kernel"` AND `"Data member of type %0 must be a field of the kernel functor"`

If we choose the latter option, I'll need to pass a flag with this information to the visitor and can then use this to determine which diagnostic to emit.

I'm open to wordsmithing all of this. Thanks!

Reply:
I think I will be okay with `"variables of type %0 must be captured directly by the kernel"` and `"data member of type %0 must be a direct member of the kernel functor"`

If two diagnostics will make for ugly code, I don't mind it being combined and they collectively referred to as `entities` maybe?

Reply:
I was taking a look at existing clang diagnostics I saw the following openCL diagnostics -
```
 "def err_opencl_addrspace_scope : Error<
  "variables in the %0 address space can only be declared in the outermost "
  "scope of a kernel function">;
```
```
def err_record_with_pointers_kernel_param : Error<
  "%select{struct|union}0 kernel parameters may not contain pointers">;
```
I also see an existing SYCL diagnostic for special classes inside Unions - 

```
def err_bad_union_kernel_param_members : Error<
  "%0 cannot be used inside a union kernel parameter">;
```

I'm leaning towards the 3rd option because its simple but conveys the problem succinctly IMO. Something like - `"%0 cannot be used inside a struct type kernel parameter".`

Technically 'used' is incorrect because unused %0 inside a used kernel parameter will also trigger this error but I don't think this is a big deal. This happens with existing union diagnostic as well. WDYT?

Reply:
```
struct B {annotated_ptr AA;};
struct A { B obj;};
struct KernelFunctor { A obj;};
```
Here field of KernelFunctor has type A. A has field of type B which in turn has field of type annotated_ptr . Is saying `annotated_ptr cannot be a field of struct kernel parameter`  confusing here ? B is not the kernel parameter, A is.

IMO saying  `annotated_ptr cannot be used inside a struct type kernel parameter` is less confusing.

Reply:
Good point.  How about `"cannot be a data member of a struct kernel parameter"`?

Reply:
Is a data member different from field? If so, that could work or maybe we can say cannot be `declared inside a struct type kernel parameter`

Reply:
> Is a data member different from field?

C++ standard uses the term data member; field is used only as bit-fields.

Reply:
Ok. Thank you for clarifying. 

I think diagnostic preference here is subjective.  I'll go with the data member version of the diagnostic you suggested. I'll modify the existing union diagnostic and combine these.

Reply:
I was thinking: we could just say "`cannot be a data member of a kernel parameter`" and avoid the "`class/struct/union type`" part of the phrasing.

Reply:
We need it for unions. streams/samplers etc are allowed to be data members of structs but not unions.